stdlib/hashtbl.ml

```diff
@@ -259,18 +252,123 @@ let to_seq tbl =
   (* state: index * next bucket to traverse *)
   let rec aux i buck () = match buck with
     | Empty ->
-        if i = Array.length tbl_data
-        then Seq.Nil
-        else aux(i+1) tbl_data.(i) ()
+      if i = Array.length tbl_data
+      then Seq.Nil
+      else aux (i+1) tbl_data.(i) ()
     | Cons {key; data; next} ->
-        Seq.Cons ((key, data), aux i next)
+      Seq.Cons ((key, data), aux i next)
   in
   aux 0 Empty
 
 let to_seq_keys m = Seq.map fst (to_seq m)
 
 let to_seq_values m = Seq.map snd (to_seq m)
 
+(* Implementation of functions dependent on key hash and equality *)
+module Impl (M: sig
+  type 'a key
+  val key_index : ('a key, 'b) t -> 'a key -> int
+  val key_equal : 'a key -> 'a key -> bool
+end) = struct
+  let key_index = M.key_index
+  let key_equal = M.key_equal
+
+  let maybe_resize h =
+    if h.size > Array.length h.data lsl 1 then resize key_index h
+
+  (* Find slot of key `k`, if it exists. Then, call continuation
+     `f h k x i prev curr`, where `h` and `k` and `x` are as passed, `i` is
+     the index of the bucket for `k`, `curr` is the first slot with key `k`
+     (or Empty if no such slot exists), and `prev` is the previous slot (or
+     Empty if `curr` is the head of the bucket). *)
+  let[@inline] find_slot h k d f =
```

Comment:
My understanding is that the `@inline` annotation does not actually guarantee that the function will be inlined; it is only a hint. To guarantee that the function will be inlined you need to add `[@inlined]` at each callsite.

Reply:
It's true that the `[@inline]` annotation doesn't guarantee that the function will be inlined, but additional `[@inlined]` annotations should have no effect.
In this case, since `find_slot` contains an inner function definition (`loop`) then only Flambda will be able to inline it. If one wants Closure to be able to inline it, then `loop` must be lifted outside `find_slot`.

Reply:
Right, thanks for the clarification.

Reply:
Note that there is no requirement to actually inline things as long as performance is generally preserved. (But we want to know that performance is also preserved without flambda, or at least to understand and discuss any performance change.)

Reply:
As I understand it `[@inlined]` on a function application raises a compilation warning if the application is not indeed inlined.

As per my measurements (native code compiler for x86 only, no bytecode or other targets), there are no extra allocations, and identical performance. But yes, without inlining there will be an extra indirect function call.